src/Abc.Zebus.Persistence.CQL/Storage/CqlStorage.cs

```diff
@@ -68,36 +65,54 @@ public Task Write(IList<MatcherEntry> entriesToPersist)
-                                                                               ToUnixMicroSeconds(rowTimestamp)));
-
-                if (shouldInvestigatePeer)
-                    insertTask = insertTask.ContinueWith(t => _log.Info($"Storage done for peer {matcherEntry.PeerId}, Type: {matcherEntry.Type}, Message Id: {matcherEntry.MessageId}, TaskResult: {t.Status}"));
-
-                insertTasks.Add(insertTask);
                 var countDelta = matcherEntry.IsAck ? -1 : 1;
                 countByPeer[matcherEntry.PeerId] = countDelta + countByPeer.GetValueOrDefault(matcherEntry.PeerId);
 
                 if (shouldInvestigatePeer)
                     _log.Info($"Count delta computed for peer {matcherEntry.PeerId}, will increment: {countDelta}");
             }
 
+            var insertTasks = new List<Task>();
+            var remaining = new SemaphoreSlim(_maxParallelInsertTasks);
+            foreach (var matcherEntry in entriesToPersist)
+            {
+                var gotSlot = remaining.Wait(TimeSpan.FromSeconds(10));
+                if (!gotSlot)
+                    _log.Warn("Could not get slot to insert in cassandra after 10 second.");
+
+                var messageDateTime = matcherEntry.MessageId.GetDateTimeForV2OrV3();
+                var rowTimestamp = matcherEntry.IsAck ? messageDateTime.AddTicks(10) : messageDateTime;
+                var boundStatement = _preparedStatement.Bind(matcherEntry.PeerId.ToString(),
+                                                             BucketIdHelper.GetBucketId(messageDateTime),
+                                                             messageDateTime.Ticks,
+                                                             matcherEntry.MessageId.Value,
+                                                             matcherEntry.IsAck,
+                                                             matcherEntry.MessageBytes,
+                                                             (int)PersistentMessagesTimeToLive.TotalSeconds,
+                                                             ToUnixMicroSeconds(rowTimestamp));
+
+                var insertTask = _dataContext.Session.ExecuteAsync(boundStatement); 
+                insertTasks.Add(insertTask);
+                insertTask.ContinueWith(t =>
+                {
+                    var shouldInvestigatePeer = _configuration.PeerIdsToInvestigate != null && _configuration.PeerIdsToInvestigate.Contains(matcherEntry.PeerId.ToString());
+                    if (shouldInvestigatePeer)
+                        _log.Info($"Storage done for peer {matcherEntry.PeerId}, Type: {matcherEntry.Type}, Message Id: {matcherEntry.MessageId}, TaskResult: {t.Status}");
+
+                    if (t.IsFaulted)
+                        _log.Error(t.Exception);
+
+                    remaining.Release();
+                });
```

Comment:
The continuation is simple and fast, it might be better to specify `TaskContinuationOptions.ExecuteSynchronously`.

Reply:
Yep no point doubling the number of tasks used.